### ISSUE TITLE
Build error using jwt-cpp-0.4.0 (latest)

### ISSUE THREAD

**Issue by duncanmmacleod** [scitokens/scitokens-cpp#33]
I'm trying to build a conda package for this, and cannot get it to build against the latest release of jwt-cpp (0.4.0). If this is an actual compatibility issue (i.e. this project needs an older version of jwt-cpp) can you please instruct me the version of jwt-cpp to use (build)?

Alternatively, it looks like this project vendors jwt-cpp at an older version in the git repo, is it possible to include that vendored copy in the tarballs to use that at build time?*

*It looks like this may be properly included in the 'release tarball' that the [spec file is referencing](https://github.com/scitokens/scitokens-cpp/blob/6b0cc40e829a24ae9d68b79a3ed098b12e592350/rpm/scitokens-cpp.spec#L13) but one of those hasn't been provided since 0.3.3.

Any help/guidance is greatly appreciated. Thanks.

Build failure:

```
[  7%] Building CXX object CMakeFiles/SciTokens.dir/src/scitokens.cpp.o
/usr/bin/c++ -DSciTokens_EXPORTS -I$SRC_DIR -I$PREFIX/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/scitokens-cpp-0.5.1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -Wall -Werror -std=c++11 -O3 -DNDEBUG -fPIC -o CMakeFiles/SciTokens.dir/src/scitokens.cpp.o -c $SRC_DIR/src/scitokens.cpp
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h: In member function ‘std::__cxx11::string scitokens::SciToken::serialize()’:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:206:47: error: no match for ‘operator=’ (operand types are ‘std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type’ {aka ‘jwt::claim’} and ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’})
         m_claims["jti"] = std::string(uuid_str);
                                               ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:6,
                 from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(const jwt::claim&)’
  class claim {
        ^~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(jwt::claim&&)’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘jwt::claim&&’
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:209:58: error: no match for ‘operator=’ (operand types are ‘std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type’ {aka ‘jwt::claim’} and ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’})
             m_claims["ver"] = std::string("scitokens:2.0");
                                                          ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:6,
                 from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(const jwt::claim&)’
  class claim {
        ^~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(jwt::claim&&)’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘jwt::claim&&’
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:212:52: error: no match for ‘operator=’ (operand types are ‘std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type’ {aka ‘jwt::claim’} and ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’})
                 m_claims["aud"] = std::string("ANY");
                                                    ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:6,
                 from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(const jwt::claim&)’
  class claim {
        ^~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(jwt::claim&&)’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘jwt::claim&&’
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:215:53: error: no match for ‘operator=’ (operand types are ‘std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type’ {aka ‘jwt::claim’} and ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’})
             m_claims["wlcg.ver"] = std::string("1.0");
                                                     ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:6,
                 from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(const jwt::claim&)’
  class claim {
        ^~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(jwt::claim&&)’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘jwt::claim&&’
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:218:80: error: no match for ‘operator=’ (operand types are ‘std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type’ {aka ‘jwt::claim’} and ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’})
                 m_claims["aud"] = std::string("https://wlcg.cern.ch/jwt/v1/any");
                                                                                ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:6,
                 from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(const jwt::claim&)’
  class claim {
        ^~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note: candidate: ‘jwt::claim& jwt::claim::operator=(jwt::claim&&)’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘jwt::claim&&’
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp: In function ‘int scitoken_set_claim_string(SciToken, const char*, const char*, char**)’:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:59:54: error: no matching function for call to ‘scitokens::SciToken::set_claim(const char*&, std::__cxx11::string)’
         real_token->set_claim(key, std::string(value));
                                                      ^
In file included from /home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens.cpp:7:
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:127:5: note: candidate: ‘void scitokens::SciToken::set_claim(const string&, const jwt::claim&)’
     set_claim(const std::string &key, const jwt::claim &value) {
     ^~~~~~~~~
/home/duncan/opt/miniconda3/conda-bld/scitokens-cpp_1606439032710/work/src/scitokens_internal.h:127:5: note:   no known conversion for argument 2 from ‘std::__cxx11::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const jwt::claim&’
make[2]: *** [CMakeFiles/SciTokens.dir/build.make:82: CMakeFiles/SciTokens.dir/src/scitokens.cpp.o] Error 1
make[2]: Leaving directory '$SRC_DIR/_build'
make[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/SciTokens.dir/all] Error 2
make[1]: Leaving directory '$SRC_DIR/_build'
make: *** [Makefile:149: all] Error 2
```

**Comment by duncanmmacleod** [comment]
After a bit of investigation, (part of) the build failure is corrected by the patch I posted in https://github.com/scitokens/scitokens-cpp/pull/16#issuecomment-734758385, but I still need to add other patches to use a released version of jwt-cpp (which are trivial). I can post them as a PR if migrating to a newer version of jwt-cpp is on the cards.

**Comment by djw8605** [comment]
Yes, please post a PR for migrating to a newer version of jwt-cpp.  We are certainly open to updating.  There is no particular reason to stay on the version we are on.

**Comment by steffen-AEI** [comment]
While attempting to build for Debian Bullseye (gcc 10.x), I found another showstopper (with jwt-cpp 0.4.0 and @duncanmmacloed's patch):
```
[  7%] Building CXX object CMakeFiles/SciTokens.dir/src/scitokens.cpp.o
/usr/bin/c++ -DSciTokens_EXPORTS -I/build/scitokens-cpp-0.5.1 -g -O2 -ffile-prefix-map=/build/scitokens-cpp-0.5.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdat
e-time -D_FORTIFY_SOURCE=2 -Wall -Werror -std=c++11 -fPIC -o CMakeFiles/SciTokens.dir/src/scitokens.cpp.o -c /build/scitokens-cpp-0.5.1/src/scitokens.cpp
In file included from /build/scitokens-cpp-0.5.1/src/scitokens.cpp:7:
/build/scitokens-cpp-0.5.1/src/scitokens_internal.h: In member function 'std::string scitokens::SciToken::serialize()':
/build/scitokens-cpp-0.5.1/src/scitokens_internal.h:206:47: error: no match for 'operator=' (operand types are 'std::unordered_map<std::__cxx11::basic_string<char>, jwt::claim>::mapped_type' {aka 'jwt::claim'} and 'std::string' {aka 'std::__cxx11::basic_string<char>'})
  206 |         m_claims["jti"] = std::string(uuid_str);
      |                                               ^
In file included from /build/scitokens-cpp-0.5.1/src/scitokens_internal.h:6,
                 from /build/scitokens-cpp-0.5.1/src/scitokens.cpp:7:
/usr/include/jwt-cpp/jwt.h:766:8: note: candidate: 'jwt::claim& jwt::claim::operator=(const jwt::claim&)'
  766 |  class claim {
      |        ^~~~~
/usr/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from 'std::string' {aka 'std::__cxx11::basic_string<char>'} to 'const jwt::claim&'
/usr/include/jwt-cpp/jwt.h:766:8: note: candidate: 'jwt::claim& jwt::claim::operator=(jwt::claim&&)'
/usr/include/jwt-cpp/jwt.h:766:8: note:   no known conversion for argument 1 from 'std::string' {aka 'std::__cxx11::basic_string<char>'} to 'jwt::claim&&'
In file included from /build/scitokens-cpp-0.5.1/src/scitokens.cpp:7:
```
repeating over and over. Is there a known cure yet?
